app/models/machine.rb

```diff
@@ -34,6 +35,9 @@ class ResurrectionExpiredError < StandardError; end
 
   has_environment default: -> { license&.environment_id }
 
+  accepts_nested_attributes_for :components, limit: 20, reject_if: :reject_associated_records_for_components
```

Comment:
We need to catch the limit error, `ActiveRecord::NestedAttributes::TooManyRecords`.